### ISSUE TITLE
Bucket not resolved when using MINIO_DOMAIN

### ISSUE THREAD

**Issue by jonhai** [minio/minio#7577]
## Expected Behavior

When the environment variable MINIO_DOMAIN is set, MinIO should support virtual-host-style requests as explained in the [documentation](https://github.com/minio/minio/tree/master/docs/config#domain). For a request with a host `mybucket.miniotest.com` and with `MINIO_DOMAIN=miniotest.com`, MinIO should pick `mybucket` as the bucket.

I've been using MinIO locally with docker, and tracked down that the last release where it works according to this expected behaviour in *RELEASE.2019-03-27T22-35-21Z*, while in the release after that, *RELEASE.2019-04-04T18-31-46Z*, it does not work anymore.

## Current Behavior

From *RELEASE.2019-04-04T18-31-46Z* onwards, the result is that when requesting `mybucket.miniotest.com` with `MINIO_DOMAIN=miniotest.com`, the bucket isn't resolved correctly.

## Steps to Reproduce (for bugs)

1. Use this `docker-compose.yml`:
````
version: "3"

services:

  minio:
    image: minio/minio:RELEASE.2019-04-04T18-31-46Z
    #image: minio/minio:RELEASE.2019-03-27T22-35-21Z # <- Works in this version
    container_name: "miniotest"
    hostname: miniotest.com
    ports:
      - "8001:9000"
    volumes:
      - "./minio-data:/data"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=miniopass"
      - "MINIO_REGION=us-east-1"
      - "MINIO_DOMAIN=miniotest.com"
    command: server /data

````
2. `docker-compose up -d`
3. Put into /etc/hosts:
````
127.0.0.1 miniotest.com mybucket.miniotest.com
````
4. Doing a simple GET request with curl gives the following:
````
$ curl http://mybucket.miniotest.com:8001/mykey

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><BucketName>mykey</BucketName><Resource>/mykey</Resource><RequestId>15985D7FA186B0D8</RequestId><HostId>c4afcac1-d638-4352-87b2-6374ce681e73</HostId></Error>
````
Here it incorrectly says BucketName=mykey when it should get the bucket from the host. Running the same with the older version *RELEASE.2019-03-27T22-35-21Z* gives:
````
$ curl http://mybucket.miniotest.com:8001/mykey

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><Key>mykey</Key><BucketName>mybucket</BucketName><Resource>/mykey</Resource><RequestId>15985DDC715A2E78</RequestId><HostId>c4afcac1-d638-4352-87b2-6374ce681e73</HostId></Error>
````
Here the bucket is resolved correctly with BucketName=mybucket.

Tried also with having the parameter `--address miniotest.com:9000` when starting the service but that did not help.

## Context

Running MinIO locally in a docker environment which involves AWS client software that only supports virtual-host-style requests.

## Regression

Seems to be regression. Last version where it works is [RELEASE.2019-03-27T22-35-21Z](https://github.com/minio/minio/releases/tag/RELEASE.2019-03-27T22-35-21Z) when the next version [RELEASE.2019-04-04T18-31-46Z](https://github.com/minio/minio/releases/tag/RELEASE.2019-04-04T18-31-46Z) breaks the functionality.

## Your Environment

* Version used (`minio version`): RELEASE.2019-04-04T18-31-46Z
* Environment name and version (e.g. nginx 1.9.1): [MinIO docker images](https://hub.docker.com/r/minio/minio), Docker Version 2.0.0.3

**Comment by nitisht** [comment]
Thanks for reporting @jonhai we'll take a look.

**Comment by maforshaw** [comment]
Same problem here. I noticed when Synology's HyperBackup started to fail, which only supports virtual-host-style requests.

Thanks @jonhai for tracking down the last working version. Downgrading to RELEASE.2019-03-27T22-35-21Z fixed the issue.

**Comment by kannappanr** [comment]
PR https://github.com/gorilla/mux/pull/383 , part of gorilla mux version v1.7.0 broke compatibility in MinIO with https://github.com/minio/minio/pull/7302 . We will be sending in a fix to handle this change.

**Comment by GreenVine** [comment]
Hi there,

I'm re-opening this issue as I am still seeing the same error with the latest version: `RELEASE.2019-05-02T19-07-09Z`.

I have set my `MINIO_DOMAIN` to `example.com` and have a bucket named `mybucket`. I'd assume that direct access to objects via `mybucket.example.com` is possible. To make the issue more obvious, I tested my deployment by using an incorrect signature to access the `testfile.txt`. With the aforementioned version:

Request: `https://mybucket.example.com/testfile.txt?X-Amz-Algorithm=SOME_INCORRECT_TOKEN`

```xml
<Error>
  <Code>SignatureDoesNotMatch</Code>
  <Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
  <BucketName>testfile.txt</BucketName>
  <Resource>/testfile.txt</Resource>
  <RequestId>159B8B79B91BDB7D</RequestId>
  <HostId>0e9f2dcf-2fc3-418f-b896-49ac3fa3d2e7</HostId>
</Error>
```

With `RELEASE.2019-03-27T22-35-21Z` as @maforshaw mentioned previously:

```xml
<Error>
  <Code>SignatureDoesNotMatch</Code>
  <Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
  <Key>testfile.txt</Key>
  <BucketName>mybucket</BucketName>
  <Resource>/testfile.txt</Resource>
  <RequestId>159B8B899D72112E</RequestId>
  <HostId>0e9f2dcf-2fc3-418f-b896-49ac3fa3d2e7</HostId>
</Error>
```

Note that `BucketName` is incorrectly captured in the latest version. Any ideas?

**Comment by harshavardhana** [comment]
> I'm re-opening this issue as I am still seeing the same error with the latest version: `RELEASE.2019-05-02T19-07-09Z`.

This issue is different @GreenVine please open a new one.

**Comment by jonhai** [comment]
Confirming that this bug still happens in *RELEASE.2019-05-02T19-07-09Z* and to me it sounds very much like the problem @GreenVine reports which now has it's own issue #7619.

Here's the what I now have the docker-compose.yml using this latest version:

````
version: "3"
  
services:

  minio:
    image: minio/minio:RELEASE.2019-05-02T19-07-09Z

    container_name: "miniotest"
    #hostname: miniotest.com
    ports:
      - "8001:9000"
    volumes:
      - "./minio-data:/data"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=miniopass"
      - "MINIO_REGION=us-east-1"
      - "MINIO_DOMAIN=miniotest.com"
      - "MINIO_HTTP_TRACE=/var/log/minio.log"
    command: server /data

````

The curl request still produces:

````
$ curl http://mybucket.miniotest.com:8001/mykey

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><BucketName>mykey</BucketName><Resource>/mykey</Resource><RequestId>159C09FBB5A36120</RequestId><HostId>41fe104b-d179-4443-8da3-373ee49d0970</HostId></Error>
````

Again with *mykey* incorrectly as the bucket name. The same happens with the *edge* version. *RELEASE.2019-03-27T22-35-21Z* is still the last version where this has worked correctly.

**Comment by kannappanr** [comment]
@jonhai I think the fix for the regression did not handle the scenario that you have described here, where the internal port is `9000`, but the external port is `8001`
Something like this will work now.

```
version: "3"
  
services:

  minio:
    image: minio/minio:RELEASE.2019-05-02T19-07-09Z

    container_name: "miniotest"
    #hostname: miniotest.com
    ports:
      - "8001:8001"
    volumes:
      - "./minio-data:/data"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=minio123"
      - "MINIO_REGION=us-east-1"
      - "MINIO_DOMAIN=test.com"
      - "MINIO_HTTP_TRACE=/var/log/minio.log"
    command: server --address :8001 /data
```
or

```
version: "3"
  
services:

  minio:
    image: minio/minio:RELEASE.2019-05-02T19-07-09Z

    container_name: "miniotest"
    #hostname: miniotest.com
    ports:
      - "9000:9000"
    volumes:
      - "./minio-data:/data"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=minio123"
      - "MINIO_REGION=us-east-1"
      - "MINIO_DOMAIN=test.com"
      - "MINIO_HTTP_TRACE=/var/log/minio.log"
    command: server /data
```


Will send in a fix that will handle the scenario you have described here. I was trying to reproduce with the same port number.

**Comment by jonhai** [comment]
@kannappanr nice, thanks!

**Comment by danielgrant** [comment]
Like @maforshaw, I too am using MinIO as a target for Synology's Hyper Backup. This has been working fine since I put it in last year, and the last known working release for me was `RELEASE.2019-04-23T23-50-36Z`. I suffered from the same problems mentioned above.

However I've now upgraded to `RELEASE.2019-05-14T23-57-45Z` and I'm still having problems. The issue described above has disappeared, but now the backups just don't run. Annoyingly, there's nothing in the MiniIO logs or the Hyper Backup logs (including `/var/log/messages`) - the backup process starts, and then just hangs indefinitely.

I've tried applying the `--compat` flag, doesn't change anything. The only thing I can currently do to get the backups working again is roll back to `RELEASE.2019-04-23T23-50-36Z`.

Interested to know if anyone else is seeing similar problems with the latest release, especially @maforshaw who is also running Hyper Backup?

Also, if it helps, the user agent that Hyper Backup is using appears to be:

```
aws-sdk-php2/2.8.31 Guzzle/3.9.3 curl/7.54.0 PHP/5.6.11
```

Perhaps that version of the AWS PHP SDK is incompatible with a recent change? But nothing I can see the changelog jumps off the page...

**Comment by GreenVine** [comment]
> Like @maforshaw, I too am using MinIO as a target for Synology's Hyper Backup. This has been working fine since I put it in last year, and the last known working release for me was `RELEASE.2019-04-23T23-50-36Z`. I suffered from the same problems mentioned above.
> 
> However I've now upgraded to `RELEASE.2019-05-14T23-57-45Z` and I'm still having problems. The issue described above has disappeared, but now the backups just don't run. Annoyingly, there's nothing in the MiniIO logs or the Hyper Backup logs (including `/var/log/messages`) - the backup process starts, and then just hangs indefinitely.
> 
> I've tried applying the `--compat` flag, doesn't change anything. The only thing I can currently do to get the backups working again is roll back to `RELEASE.2019-04-23T23-50-36Z`.
> 
> Interested to know if anyone else is seeing similar problems with the latest release, especially @maforshaw who is also running Hyper Backup?
> 
> Also, if it helps, the user agent that Hyper Backup is using appears to be:
> 
> ```
> aws-sdk-php2/2.8.31 Guzzle/3.9.3 curl/7.54.0 PHP/5.6.11
> ```
> 
> Perhaps that version of the AWS PHP SDK is incompatible with a recent change? But nothing I can see the changelog jumps off the page...

@danielgrant Did you try to re-setup your Hyper Backup connection?

My current symptoms with Hyper Backup are when creating custom S3 destination, it prompts `The operation failed. Please log in to DSM again and retry.`. From `/var/log/messages` I can see the following logs:

```
2019-05-15T19:41:24+10:00 SynologyHostname synoscgi_SYNO.Backup.Target_1_get_candidate_dir[12024]: (12024) [err] error_mapping.cpp:32 listDirEx:1161: failed, {"aws_error_code":"AuthorizationHeaderMalformed","aws_error_type":"client","error_class":"Aws\\S3\\Exception\\S3Exception","error_message":"The authorization header is malformed; the region is wrong; expecting 'us-east-1'.","http_status_code":400,"success":false}
2019-05-15T19:41:24+10:00 SynologyHostname synoscgi_SYNO.Backup.Target_1_get_candidate_dir[12024]: [12024]cloud_tm.cpp:336 Error: list dir failed: [3]
```

My region settings in the `config.json` is to a non-empty string other than `us-east-1`. Minio release is `RELEASE.2019-05-14T23-57-45Z`.

**Comment by danielgrant** [comment]
@GreenVine I'm not getting any messages like that in `/var/log/messages`. Interestingly Hyper Backup seems capable of interacting with MinIO to an extent, for example I am able to browse the backups on MinIO via the Hyper Backup UI, I'm able to view the total size of the remote backup via the Hyper Backup UI, the Hyper Backup UI shows the backup destination as "Online", and when observing the access log for the NGINX proxy I have sitting in front of it I can see a small number of initial requests from Hyper Backup to MinIO (all of which return a 200), followed heartbeat every 60 seconds that is hitting a certain file in MinIO that also returns a 200.

But aside from that, the backup never progresses. Hyper Backup just hangs on "Processing data", `/var/log/messages` doesn't output anything else, the `journalctl` for MinIO produces nothing, and the only traffic the NGINX logs show is the heartbeat every 60 seconds... :man_shrugging: 

It would be useful if I could increase the verbosity of MinIO's logging to see if that reveals anything but I can't see any documentation on how to do that - there's an HTTP trace though so I might have to try that.

**Comment by GreenVine** [comment]
> It would be useful if I could increase the verbosity of MinIO's logging to see if that reveals anything but I can't see any documentation on how to do that - there's an HTTP trace though so I might have to try that.

@danielgrant Definitely try to enable HTTP trace via `MINIO_HTTP_TRACE` environment variable. I did see lots of 400 and 403 requests in my case.

I will stick with the earlier release for now.

**Comment by kannappanr** [comment]
@danielgrant and @GreenVine Can you please open a new issue. This is probably a new issue though you are seeing it on similar setups.

**Comment by maforshaw** [comment]
>  Interested to know if anyone else is seeing similar problems with the latest release, especially @maforshaw who is also running Hyper Backup?

@danielgrant I just upgraded to the latest minio server to test and Hyper Backup appears to work fine. The backup runs and completes.

As suggested by @GreenVine I recommend you set the `MINIO_HTTP_TRACE` environment variable to a log file to see what is going on. It will log every request and response.

I don't know if it matters but the difference in my setup is I'm not using a reverse proxy. Hyper Backup is going directly to the minio port I specified.
